### ISSUE TITLE
Uninformative type-parameter constraint error with recursive data

### ISSUE THREAD

**Issue by jeffreymorlan** [Microsoft/TypeScript#7269]
``` ts
interface MyInterface {
    children: MyInterface[];
    someOtherProperty: any;
}
function foo<T extends MyInterface>() {}

class MyClass {
    children: MyClass[];
}
foo<MyClass>(); // Error
```

TS 1.7.5's message helpfully points out the missing property :)

```
Type 'MyClass' does not satisfy the constraint 'MyInterface'.
  Property 'someOtherProperty' is missing in type 'MyClass'.
```

TS 1.8.2's error message just amounts to "it's wrong because it's wrong" :(

```
Type 'MyClass' does not satisfy the constraint 'MyInterface'.
  Types of property 'children' are incompatible.
    Type 'MyClass[]' is not assignable to type 'MyInterface[]'.
      Type 'MyClass' is not assignable to type 'MyInterface'.
```

**Comment by sandersn** [comment]
The error elaboration usually just chooses the first incorrect property. I'm not sure why we skipped `children` in 1.7. I'll go find out.

**Comment by sandersn** [comment]
The change that introduces this error was in PR #5761 to fix #5712. In the new baselines you can already see this problem showing up with `IPromise<any>` vs `Promise<any>`.

**Comment by sandersn** [comment]
The fix is to revert #5761. Some other code change must have fixed the source of that #5712's crash in the intervening months. Fix is up at #8861.
